### ISSUE TITLE
Resource.all() can't be iterated if empty

### ISSUE THREAD

**Issue by melinath** [paypal/PayPal-Python-SDK#170]
### General Information

- Mode (Sandbox/Live): Both, presumably
- PayPal-Debug-ID(s) (from any logs): N/A
- Version of Python used: 3.5
- SDK Version: 1.11.7

### Issue Description

> Please include as many details (logs, steps to reproduce, frameworks/dependencies) as you can to help us reproduce this issue faster.

I'm trying to create a web experience object "singleton" - i.e. I want to have a web experience object and create it if it doesn't exist. Here's the code I'm using:

```python
def get_web_profile():
    profiles = paypalrestsdk.WebProfile.all()
    profile = None
    for p in profiles:
        if p.name == WEB_PROFILE_NAME:
            profile = p
            break

    if not profile:
        profile = create_web_profile()

    return profile
```

Leaving aside that it's ridiculous I have to do this - this code doesn't work (though it should). Specifically, it fails when I try to iterate over the (empty, because I haven't created any yet) profiles object. More specifically, I get the following error:

```
File <redacted> in get_web_profile
    for p in profiles:
  File "<redacted>/paypalrestsdk/resource.py", line 80, in __getitem__
    return self.__data__[key]
```

So basically, iteration is implemented with `__getitem__`. I suspect this may work when the list is populated, but if the list is empty, then `super(Resource, self).__setattr__('__data__', {})` (currently line 16) sets data to an empty dict, which isn't indexable.

**Comment by melinath** [comment]
The obvious workaround (checking whether `profiles` is truthy before trying to iterate over it) doesn't work because despite being empty, `profiles` _is_ truthy. :-/ I guess I'll have to either check __data__ directly or do a try/except.

**Comment by jaypatel512** [comment]
Thanks @melinath for detailed issue. I am not a python expert as such, and I have asked my team mates to look into this (who indeed is python guru). 

Do you think of any change that we could do to make it easier for all cases ? Probably an empty iterable, that will not throw above exception.

**Comment by melinath** [comment]
@jaypatel512 ended up having to give up on the rest framework and go classic because of backwards-compatibility with existing subscriptions. But. Basically you need to implement the following items:

* [`__nonzero__`](https://docs.python.org/2/reference/datamodel.html#object.__nonzero__) so that a resource can be empty or not based on its internal state. ([`__bool__`](https://docs.python.org/3/reference/datamodel.html#object.__bool__) in python 3.)
* [`__iter__`](https://docs.python.org/2/reference/datamodel.html#object.__iter__) so that resources that are iterable can be actually iterated over instead of falling back on `__getitem__`.
* [`__getitem__`](https://docs.python.org/2/reference/datamodel.html#object.__getitem__) implementation altered so that it behaves appropriately depending on whether the resource is, internally, a mapping object or an indexed object. (This could obviate the need for implementing `__iter__`)

**Comment by sucithrashanu** [comment]
@melinath 

Currently we do not have any stipulated timeline for the v2.0.0 release, but it may be available in near future.

Please Check PayPal Community  (https://www.paypal-community.com )and  MTS ( https://paypal.com/mts )
Thanks!
